### PR TITLE
WW-4744: AnnotationUtils supports non-public methods

### PR DESCRIPTION
**Pull request by yasserzamani**
With these changes, AnnotationUtils also has equipped with an internal cache. Mainly, getAnnotatedMethods and all it's usages have improved and equipped with enough unit tests (also all findAnnotation usages have equipped).

NOTES:
To avoid reinventing the wheel, I copied from Spring framework, So, I copied their license and authors as well where needed :)

### CODE REVIEW

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java [hunk 1]
```diff
@@ -37,13 +40,25 @@
  * @author Rainer Hermanns
  * @author Zsolt Szasz, zsolt at lorecraft dot com
  * @author Dan Oxlade, dan d0t oxlade at gmail d0t c0m
+ * @author Rob Harrop
+ * @author Juergen Hoeller
+ * @author Sam Brannen
+ * @author Mark Fisher
+ * @author Chris Beams
+ * @author Phillip Webb
```
Comment (lukaszlenart) [inline core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: I'm not sure if adding those authors here makes any sense, in most cases we (at ASF) trying to remove the `@author` tag also asking those users about this class can confuse them.

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java [hunk 2]
```diff
@@ -150,25 +166,32 @@ public static void addAllInterfaces(Class clazz, List<Class> allInterfaces) {
...
+            while (result == null) {
+                clazz = clazz.getSuperclass();
+                if (clazz == null || clazz.equals(Object.class)) {
+                    break;
+                }
+                try {
+                    Method equivalentMethod = clazz.getDeclaredMethod(method.getName(), method.getParameterTypes());
+                    result = getAnnotation(equivalentMethod, annotationType);
+                } catch (NoSuchMethodException ex) {
+                    // No equivalent method found
```
Comment (lukaszlenart) [inline core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: no need to LOG this exception?

**Comment by lukaszlenart** [comment]
Looks good, just have some doubts about putting `@author` tag

**Comment by cnenning** [comment]
> Should I ask legal [1] about that? What do you think?

That's the class that has been copied from spring. Yes, I think it's better to ask how to handle this.

**Comment by lukaszlenart** [comment]
> That's the class that has been copied from spring

I meant, coping classes is ok if the license allows that (Spring uses AL 2.0) but my questions was about adding authors to some other class (not copied from Spring), maybe let's allow Yasser explain why he did that.

**Comment by aleksandr-m** [comment]
Do we need custom annotation utility at all? Can't we just use existing libraries e.g. Apache Commons [MethodUtils](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/reflect/MethodUtils.html)?

**Comment by yasserzamani** [comment]
> Yasser explain why he did that

Thank you for your review.

`AnnotationUtils` is not a new class but to avoid reinventing the wheel, I copied and merged some codes from Spring into that. While Spring has `Copyright 2002-2014 the original author or authors.` in his License header, I was not sure what to do with `@author`s :(

Currently, I am searching Internet for same to see what people do. Finally, I also may ask Spring.

**Comment by lukaszlenart** [comment]
We must rather ask ASF legal body, you can post an issue in JIRA (see above)

**Comment by yasserzamani** [comment]
> We must rather ask ASF legal body, you can post an issue in JIRA (see above)

Thank you. As I could not find similar issue, I created [Merge others codes having same license but different copyrights](https://issues.apache.org/jira/browse/LEGAL-295).

File: core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java (hunk 2 above)
Comment (yasserzamani) [inline core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java]: No, it is common to not found the method in one of supercleasses during our hierarchically search.

**Comment by yasserzamani** [comment]
I also wait for ASF LEGAL part's answer but after some study, it seems I was not allowed to do such work 😞  If so, I think I have to write my own utils at where @aleksandr-m mentioned; re-inventing the wheel :(

Below are references; what do you think, please?

References:
[1] [https://www.apache.org/licenses/LICENSE-2.0](https://www.apache.org/licenses/LICENSE-2.0)

> 4. Redistribution.
> c. You must retain, in the Source form of any Derivative Works that You distribute, all copyright, patent, trademark, and attribution notices from the Source form of the Work, excluding those notices that do not pertain to any part of the Derivative Works; and

[2] [http://www.apache.org/legal/src-headers.html](http://www.apache.org/legal/src-headers.html)

> TREATMENT OF THIRD-PARTY WORKS
> 1. Do not modify or remove any copyright notices or licenses within third-party works.

[3] [http://www.apache.org/dev/apply-license.html](http://www.apache.org/dev/apply-license.html)

> CAN/SHOULD INDIVIDUAL COMMITTERS ADDED COPYRIGHT STATEMENTS TO THE NOTICE OR SOURCE CODE FILES?
> 
> No.
> 
> Though committers retain copyright, Apache asks that they do not add copyright statements. See the policy for more details.

**Comment by yasserzamani** [comment]
> Do we need custom annotation utility at all? Can't we just use existing libraries e.g. Apache Commons MethodUtils?

Thank you @aleksandr-m !

I am working on it. I started by creating [LANG-1317](https://issues.apache.org/jira/browse/LANG-1317)

**Comment by yasserzamani** [comment]
✅ All copied Spring's code including what I was imported in previously merged [PR#117](https://github.com/apache/struts/pull/117) removed 👌 

After merge of [LANG-1317 PR](https://github.com/apache/commons-lang/pull/261) and S2's upgrade to use commons lang 3.6, I'll remove deprecated `getAnnotatedMethods` and `findAnnotation` to Apache Commons Lang 3.6 👌

**Comment by lukaszlenart** [comment]
So as I understand this is ready to be merged with a future notice to use Commons Lang 3.6 when available?

**Comment by yasserzamani** [comment]
@lukaszlenart , Yes in my opinion, Thank you!

**Comment by lukaszlenart** [comment]
Cool, I will review this tomorrow just to double check and LGTM! Great work!

**Comment by lukaszlenart** [comment]
Great work, LGTM! 👍

**Pull request merged**